### ISSUE TITLE
With "ao=pipewire" changing ao-volume doesn't match mixer reported value

### ISSUE THREAD

**Issue by xkero** [mpv-player/mpv#11065]
Important Information
mpv version: 0.35.0
Linux Distribution and Version: Archlinux
Source of the mpv binary: Archlinux community repo
pipewire version: 0.3.63

So 0% and 100% match, but values in-between don't. So setting volume to 2% in mpv via ao-volume results in the volume showing as 27% in both pavucontrol and kde's volume mixer, 12% in mpv is 50% in both mixers, 74% in mpv is 91% in both mixers, etc. Setting `--ao=pulse` fixes this, so e.g. 50% in mpv is 50% in the mixer. I looked through the man page for mpv thinking maybe there was some setting for how volume is scaled maybe, but couldn't find anything. If it isn't a bug I find it weird that pulseaudio is fine, but pipewire is different.

**Comment by t-8ch** [comment]
I'm sure this worked before, I'll try to reproduce it later.
How many audio channels does your media have?
How do changes in pavucontrol reflect in mpv?

**Comment by xkero** [comment]
> How many audio channels does your media have?
Stereo

> How do changes in pavucontrol reflect in mpv?
Same way, so setting it to 0% in mpv, then moving it to 52% in pavucontrol, and finally pressing the key to lower ao-volume in mpv the osd says 12%, but the mixer's both say 50% now.

It seems to act like mpv is using a different scale or formula to work out the volume percentage, because as mentioned 0% and 100% still match up.

**Comment by t-8ch** [comment]
So ao_pipewire correctly sets the volume to pipewire
But while pipewire uses linear volumes,  pulse uses cubic ones. pipewire-pulse converts the volume for clients so the effective value matches.

But it is confusing indeed.

Originally ao_pipewire used cubic volume but that got changed in #10356 by the author of pipewire himself .

**Comment by winex** [comment]
_tried rebinding controls to `ao-volume` for visual feedback in audio mixer (`kmix`) when running multiple instances of `mpv` and stuck with this bug!_

omg, it's been **4 years** already and nobody commented:
- browsers/javascript/html5/who-did-that: linear volume multiplier 😠
- `pipewire`: linear _(is this still true in 2026?)_
- `pulse`: **cubic**
- `kmix`: **cubic** (might be using `pulse` layer, need to check repo)
- `wpctl`: **cubic** for  `set-volume` / `get-volume`
- 99.42% of stupid apps: linear
- `mpv`: ... was... **cubic**!

@wtay: `make mpv use linear volumes because of pipewire`... **why on earth** `mpv` should do the same if it's unusable for humans?!

instead, `pipewire` should probably have used `powf(vol, 3.0f);  // or v*v*v;` _(to process signal, ofc, but keep "slider" value)_ in the first place, so nobody ever does stupid linear volumes again!

### ...
terminology, just for the 1001th time:
- volume slider IS linear: `0 .. 100 %`, which means `0.0 .. 1.0`, but multiplying signal by this value is idiotic
- audible volume/loudness/SPL IS **logarithmic**: `-inf .. 0 dB`
- volume sliders could be logarithmically mapped to **either**: `-30, -40 or -60 dB`, _but
  what if someone wants, for ex, -90dB mapping for `PCM_S24LE`? => patch again and recompile..._
- `<N> dB`-mapped volume sliders at minimal values never fall to `-inf` and it's the second reason why we always have `MUTE` button nearby...

... so mapping volume slider to **cubic** curve is:
- way`**3` easier to implement
- enough to attenuate `PCM_S16LE` to silence: `-1 .. 1` seems to be `-45dB`, while **`1%` linear is only `-20dB`!**
- feels natural, very smooth and **usable in its full range!**
- cubic-root `cbrt(vol_cubic);  // or powf(vol_cubic, 1/3.0f);` is used to set the slider back for humans:
  `cbrt(0.000125f) == 5%`

### tldr;
just revert https://github.com/mpv-player/mpv/commit/c7b17beaf1f3799bf5a1fd3ebe21c9a3e78d48da from #10356

**Comment by llyyr** [comment]
> mpv: ... was... cubic!

mpv is cubic. You're adjusting ao-volume property, which is entirely dependent on the audio server. Please read the property's documentation:

> What this option does, or how the value is interpreted depends on the API. For example, on ALSA this usually changes system-wide audio volume on a linear curve, while with PulseAudio this controls per-application volume on a cubic curve.

**Comment by winex** [comment]
> mpv is cubic. You're adjusting ao-volume property, which is entirely dependent on the audio server. Please read the property's documentation:

with `--ao=pipewire`, property `ao-volume` controls application/stream volume, it never touches anything related to ALSA.
_specifically `channelVolumes` until `--pipewire-mode=global`, but `kmix` doesn't support global-volume property, yet... time for another PR_

### fix docs
```diff
-while with PulseAudio ...
+while with PipeWire and PulseAudio ...
```
this: https://github.com/winex/mpv/commit/eb31385d44d849930814e02e5c3ae139020595cd

### test it:
using `pipewire` as audio server, `pipewire-pulse` compat layer and `wireplumber`:
```
$ mpv --ao=pulse --no-terminal --input-commands='set ao-volume 50' test.wav &
$ wpctl status | grep -i mpv
$ wpctl get-volume $MPV_ID
Volume: 0.50
```
but (this is **wrong**):
```
$ mpv --ao=pipewire --no-terminal --input-commands='set ao-volume 50' test.wav &
$ wpctl status | grep -i mpv
$ wpctl get-volume $MPV_ID
Volume: 0.79
```

**Comment by mk-fg** [comment]
I've also installed fresh pipewire + wireplumber with all-default config today, and noticed this issue, with a larger difference in volumes towards lower end of the range.

Shell-loop to plot mpv volume vs wireplumber/pipewire numbers:
```
file=test.wav; for vol in 1 11 21 31 41 51
do mpv --no-config --no-terminal --ao=pipewire --input-commands="run sleep 0.1,set ao-volume $vol" "$file" & pid=$! \
&& sleep 1 && vol_pw=$(wpctl get-volume $(wpctl status | awk '/mpv/ {n=$1+0} END {print n}')) \
&& kill $pid && echo " -- mpv=$vol pw=$vol_pw" || break; done |& grep mpv=
```

This prints the following relationship:
```
 -- mpv=1 pw=Volume: 0.22
 -- mpv=11 pw=Volume: 0.48
 -- mpv=21 pw=Volume: 0.59
 -- mpv=31 pw=Volume: 0.68
 -- mpv=41 pw=Volume: 0.74
 -- mpv=51 pw=Volume: 0.80
```

Not sure if it works out to a cubic curve, but wanted to point out that this is much more of an issue in <50% pipewire volume range, and if you need volume from there, usable ao-volume range in mpv is 0-10% (if not 0-1% to get <20%).
Quite reminiscent of [loudness war], when you already have to keep volume low to not go deaf, and compounding that problem.

Arch Linux OS (current as of 2026-04-11), mpv 0.41.0, pipewire 1.6.3, wireplumber 0.5.14

[loudness war]: https://en.wikipedia.org/wiki/Loudness_war
